---
title: "Course Introduction"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_document

---

```{r eval=TRUE,echo=FALSE,results='asis'}
BiocStyle::markdown()
```


# Welcome!



# Who's who 
![](http://www.cruk.cam.ac.uk/sites/default/files/styles/leading/public/basic-pages/CRUK-CI-external-autumn.jpg?itok=849P8Q53)    

We're based at [Cancer Research Uk Cambridge Institute](http://www.cruk.cam.ac.uk/), about 2 miles south of here

![](../images/whos-who.png)



## Top Row
- Mark
    + Bioinformatics Training Coordinator, Bioinformatics Core
- Andy
    + Senior Research Associate, Tavare lab (Statistics and Computational Biology )
- Oscar
    + Senior Research Associate, Caldas lab (Breast Cancer Genomics)
    
## Second Row

- Geoff
    + Senior Research Associate, Markowetz (Systems Genetics of Cancer) and Brenton (Functional genomics of ovarian cancer)
- Matt
    + Bioinformatics Core Head
- Gabriella
    + Bioinformatics Training Facility
- Paul
    + Bioinformatics Training Facility
    
Huge thanks too to Helen Anderson (CRUK) and Cathy Hemmings (Genetics) for admin support

# Who are you?

```{r eval=FALSE,echo=FALSE}
library(dplyr)
library(xkcd)
ppl <- tbl_df(read.csv("../JUNE 2016 LIST.csv")) 

data <- count(ppl, Institute)
data$xmin <- 1:10
data$xmax <- data$xmin+1
data$ymin = 1
data$ymax = data$count

mapping <- aes(xmin = xmin, xmax=xmax, ymin=ymin,ymax=ymax)
xrange <- range(min(data$xmin),max(data$xmax))
yrange <- range(min(data$ymin),max(data$ymax))

p <- ggplot() + xkcdrect(mapping, data) + xkcdaxis(xrange,yrange)
 
```


```{r echo=FALSE, message=FALSE,fig.width=12}
library(dplyr)
library(ggplot2)
library(gridExtra)
ppl <- tbl_df(read.csv("../JUNE 2016 LIST.csv"))
p1 <- ggplot(ppl, aes(x = Institute)) + geom_bar(fill=rgb(29,0,150,maxColorValue=255)) + coord_flip()
p2 <- ggplot(ppl, aes(x=Position)) + geom_bar(fill=rgb(236,0,140,maxColorValue=255)) + coord_flip()
p3 <- ggplot(ppl, aes(x=Gender,fill=Gender)) + geom_bar() + scale_fill_manual(values=c(rgb(29,0,150,maxColorValue=255),rgb(236,0,140,maxColorValue=255)))
grid.arrange(p1,p2,p3,ncol=3)
```


##Admin


# About the Course

- We will tell about '*best practice*' tools that we use in daily work as Bioinformaticians
- You will (probably) not come away being an expert
- We cannot teach you everything about NGS data
    + plus, it is a fast-moving field
- Whole genome and exome DNA-seq only
    + much of the initial processing is the same for other assays
    + see our [previous summer school](http://bioinformatics-core-shared-training.github.io/cruk-bioinf-sschool/) for details of RNA- and ChIP-seq
- However, we hope that you will
    + Understand how your data are processed
    + Be able to explore your data - no programming required
    + Understand the issues
    + Communicate more effectively with the Bioinformaticians dealing with your data
    + Increase confidence with R and Bioconductor
    + Know where to get help
    + Be able to explore new technologies, methods, tools as they come out
    + Meet your colleagues from other CRUK institutes



# Course Outline

## Day 1

- Recap of R
- Overview of NGS data
- Hands-on experience with NGS data


## Day 2

- Theory of SNV-calling
- How SNV calls are represented
- Copy-number analysis

## Day 3


- Issues with Somatic calling
- Hands-on practice assessing Somatic calls
- Annotating and prioritsing calls

## Day 4

- Structural Variants

## Day 5

- Dealing with collections of genomes


## Disclaimer

![](../images/fisher.jpg)

> To consult the statistician after an experiment is finished is often merely to ask him to conduct a post mortem examination. He can perhaps say what the experiment died of.‚Äù. R.A. Fisher, 1938

If you haven't designed your experiment properly, then all the Bioinformatics we teach you won't help: Consult with your local statistician - preferably **not the day before your grant is due!!!!**

<iframe width="420" height="315" src="https://www.youtube.com/embed/Hz1fyhVOjr4" frameborder="0" allowfullscreen></iframe>

- We have some [materials](http://bioinformatics-core-shared-training.github.io/experimental-design/) you can look at

## Also bear in mind..

- No "*one-size fits all*" solution...

    + Every dataset will have it's own quirks and artefacts
    + You'll need to become familar with how to inspect your data
    +     
    
    
# About the data

We will use publically-available datasets that you can take away with you at the end of the course

## HCC1143

A breast-cancer cell-line on which we have data available and has been used as part of a "*benchmarking*" dataset to evaluate different variant-callers

![](http://www.pawefish.path.cam.ac.uk/BreastLineImages/Hcc1143m24w.gif)

- http://www.pawefish.path.cam.ac.uk/BreastCellLineDescriptions/HCC1143.html
- https://cansar.icr.ac.uk/cansar/cell-lines/HCC1143/copy_number_variation/
- https://cghub.ucsc.edu/datasets/benchmark_download.html



# About the software

