---
title: "FurtherSNVThings"
author: "Andy Lynch"
date: "21 July 2016"
output: html_document
---

First load the libraries that we will require:

```{r}
library(SomaticSignatures)
library(BSgenome.Hsapiens.UCSC.hg19)
library(deconstructSigs)
```


Now read in the vcf file, and keep just the autosomes and sex chromosomes (which we have to rename to match the bioconductor reference genome).

```{r}
vcf<-readVcfAsVRanges("HCC1143_vs_HCC1143_BL.flagged.muts.vcf","hg19",use.names=T)
vcf<-keepSeqlevels(vcf,c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","X","Y"))
seqlevels(vcf)<-c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY")
```

The VRanges object contains two rows for each SNV (one for the control and one for the normal), so we just take the bottom half. We will also remove anything failing the soft filters.

```{r}
vcf<-vcf[92284:184566]
vcf<-vcf[apply(vcf@softFilterMatrix,1,prod)==1]
```

We can obtain the mutational context for each SNV quite conveniently

```{r}
mc<-mutationContext(vcf,BSgenome.Hsapiens.UCSC.hg19)
```

and can now reformat the contexts to match the signature data, and tally up the numbers of each.

```{r}
SMT<-paste(
  subseq(elementMetadata(mc)$context,1,1),
  rep("[",length(mc)),
  subseq(elementMetadata(mc)$alteration,1,1),
  rep(">",length(mc)),
  subseq(elementMetadata(mc)$alteration,2,2),
  rep("]",length(mc)),
  subseq(elementMetadata(mc)$context,3,3),
  sep="")

datatemplate<-randomly.generated.tumors[1,]
for(j in 1:96){datatemplate[1,j]<-sum((SMT==colnames(signatures.cosmic)[j]))}
```

Before going further we can have a look at the mutation numbers. 

```{r}
boxplot(t(as.matrix(datatemplate[1,]))~sapply(names(datatemplate),substr,2,6))
```

Having seen that transitions are dominating, we can pull out the signatures.

```{r}
sigs<-whichSignatures(tumor.ref =datatemplate, 
                                signatures.ref = signatures.cosmic, 
                                sample.id = 1, contexts.needed = TRUE)

sigs$weight

```

Which signature dominates? Compare to the list at http://cancer.sanger.ac.uk/cosmic/signatures and decide whether this is as expected.


## MNVs

We will now take a quick look at the multiple nucleotide variants in the vcf. These manifest as consecutive bases represented in consecutive rows of the file.

```{r}
MNVindex<-which(diff(ranges(mc)@start)==1)
MNVindex<-MNVindex[-(which(diff(MNVindex)==1)+1)]
```

We can extract the sequences of each. 

```{r}
MNV<-rep("",275)
for(i in 1:275){
mnvlength<-max(which(ranges(mc)@start[MNVindex[i]+1:100]-ranges(mc)@start[MNVindex[i]]==1:100))  
MNV[i]<-paste(paste(mc@ref[MNVindex[i]+(0:mnvlength)],collapse=""),">",paste(mc@alt[MNVindex[i]+(0:mnvlength)],collapse=""),collapse="",sep="")
}

```

Note that we haven't collapsed equivalent mutations (e.g. TT>AA and AA>TT) into one entry for this object. Which MNVs are most prevalent? Of what might they be indicative?

```{r}
sort(table(MNV))
```

One MNV appears to be TGTG>CCCT. It is located at chr5:134119386-134119389. Look at this area in IGV and decide whether you believe it.

